Split boilerplate test

abadger opened the pull request:
##### SUMMARY
Split the boilerplate test into a separate one for ```__metaclass__ = type``` and our ```from __future__ import```s.

be explicit about which files we're skipping but want to fix eventually so that they can be removed one-by-one.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
The ```__metaclass__ = type``` change shouldn't break anything so we want to be able to fix these en masse.  Need to separate that from the ```from __future__``` check because the ```from __future__``` check requires some light auditing to make sure that the change is safe.

abadger committed:
Split the metaclass == type and from future boilerplate code.

The metaclass boilerplate is safe to apply en masse.  The future import
boilerplate needs code to be inspected to be sure that there aren't any
py2isms that need to be fixed.  Split these two checks so that we can
fix them independently

Be explicit about which files are grandfathered so we can fix them up one by one

mattclay commented on docs/docsite/rst/dev_guide/testing/sanity/future-import-boilerplate.rst:
```diff
@@ -1,10 +1,8 @@
-Sanity Tests » boilerplate
```
Keep the original `boilerplate` test docs even though the test is gone. That way links to the docs from old PRs will continue to function.

abadger replied:
Done.

mattclay commented on docs/docsite/rst/dev_guide/testing/sanity/type-boilerplate.rst:
```diff
@@ -0,0 +1,8 @@
+Sanity Tests » type boilerplate
+===============================
+
+Most Python files should include the following boilerplate:
```
It would be helpful to mention that this must come after any `__future__` import, but before any class definitions.

I'd also briefly explain why it's important.

abadger replied:
type-boilerplate now starts with:

> Most Python files should include the following boilerplate at the top of the file, right after the
comment header and ``from __future__ import``:

and an explanation of what that does:
https://github.com/ansible/ansible/pull/58641/files#diff-e27c8e1cc94b9db801f598584e2d4a58

mattclay commented on docs/docsite/rst/dev_guide/testing/sanity/future-import-boilerplate.rst:
```diff
@@ -1,10 +1,8 @@
+Sanity Tests » from __future__ boilerplate
+==========================================
 
 Most Python files should include the following boilerplate:
```
It would be helpful to mention that this must be at the beginning of the file.

I'd also briefly explain why it's important.

abadger replied:
Th ```__future__``` boilerplate documentation now starts with
> Most Python files should include the following boilerplate at the top of the file, right after the
comment header:

and contains an explanation of what each of the ```__future__``` impotrs mean https://github.com/ansible/ansible/pull/58641/files#diff-d0c964e228f5c86ed5f7b38621559b5e

mattclay commented on test/sanity/code-smell/type-boilerplate.py:
```diff
@@ -0,0 +1,322 @@
+        'hacking/',
+        'lib/ansible/module_utils/csharp/',
+        'lib/ansible/module_utils/powershell/',
+        'lib/ansible/module_utils/aws/__init__.py',
```
Move filenames out of `prune` into `skip` -- just separate them from the other entries with a comment.

abadger replied:
I can do that as long as we don't care that we then have to read and check for the boilerplate on every file in the skip_list.  (grandfathered currently has to be searched in order to tell people when to remove a file from the grandfathered list).

mattclay replied:
As long as the files in the `prune` list are going to be on a separate `grandfathered` list that will be fine. It just didn't seem right to have files in the `prune` list when it was intended only for directories.

abadger committed:
Rename python files in hacking/ directory to have .py suffix

ansible-test only passes files which have the .py suffix for sanity
tests on python files.  This change will allow sanity tests to run on
the Python files in hacking/

* Rename test-module to test-module.py
* Symlink test-module for backwards compat since end users may be using
  test-module
* Fix test-module sanity errors that are now triggered
* Rename ansible_profile to ansible-profile.py
* Rename build-ansible

mattclay commented on docs/docsite/rst/dev_guide/testing/sanity/future-import-boilerplate.rst:
```diff
@@ -0,0 +1,51 @@
+
+    from __future__ import (absolute_import, division, print_function)
+
+This uses Python3 semantics for absolute vs relative imports, division, and print.  By doing this,
```
Docs should refer to `Python 2` and `Python 3` instead of `Python2` and `Python3`.

abadger replied:
Fixed

mattclay commented on test/sanity/code-smell/boilerplate.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "disabled": true
+}
```
Deleted tests should only have documentation left behind. No stub test or config file is required.

The docs will need to be marked as an orphan though:

```
:orphan:
```

abadger replied:
Done.

mattclay commented on test/sanity/code-smell/type-boilerplate.py:
```diff
@@ -0,0 +1,1004 @@
+#!/usr/bin/env python
```
This would be better named `metaclass-boilerplate` instead of `type-boilerplate`.

abadger replied:
Done

mattclay commented on docs/docsite/rst/dev_guide/testing/sanity/metaclass-boilerplate.rst:
```diff
@@ -0,0 +1,23 @@
+Sanity Tests » metaclass boilerplate
```
A minor tweak for consistency with the title of the future boilerplate page title:

```suggestion
Sanity Tests » __metaclass__ boilerplate
```

abadger replied:
Hmm... I forgot that the title doesn't need to match up with the test name.... That being the case I'll change this to __metaclass__ = type boulerplate.

abadger replied:
Fixed and updated grandfathered list pushed.  If it passes, I'll merge it and let slack know that everything is stale.

abadger commented:
I merged it as two commits in case you want us to backport the test split but not the hacking*.py renaming.
